### ISSUE TITLE
cmd/compile: multiple compiler crashes and failures with -G=3

### ISSUE THREAD

**Issue by mdempsky** [golang/go#46704]
# Background

My main motivation for unified IR is to reduce complexity and duplicated code, under the expectation that this will reduce the frequency of compiler issues and ultimately reduce the engineering cost of having a production-quality generics implementation ready to ship for Go 1.18.

After getting unified IR to its first milestone of feature completion and review readiness, I decided to do a first round of aggressive testing to help empirically test my hypothesis. Up to this point, I'd mostly just relied on existing tests and `toolstash -cmp`. The only additional test case I'd written specifically for unified IR was [nested.go](https://go-review.googlesource.com/c/go/+/324573/11/test/typeparam/nested.go), to test type shadowing and generic types declared within generic functions (as those are intentionally not supported by -G=3 mode).

# Methodology

I set a goal of finding 10 issues. I wrote test cases manually and avoided local type definitions (again, because -G=3 intentionally doesn't support them). I skimmed through the -G=3 code for inspiration (e.g., that's how I found the `comparable` failure in case 8), but largely focused on variations of cases that I know from experience have been a sore point for the Go compiler.

I did not specifically set out looking for compiler crashes. I was focused on broad testing of functionality, not deep testing, so whenever I found an issue I just minimized it, and then moved on to trying other ideas. I did not explore if any of the compiler crashes could be turned into miscompilation errors.

# Results

I found 10 issues within 3 hours. They're available at https://go-review.googlesource.com/c/go/+/326169 as test/run.go compatible tests. @cuonglm also reported an 11th in the comments.

All 11 issues affect -G=3 mode. Only 1 of them affects unified IR (see below for details). I think these findings support my hypothesis that unified IR's design reduces the frequency of compiler issues.

9 of the issues are the compiler crashing on valid code. 1 issue is the compiler accepting invalid code. 1 issue is the compiler crashing on invalid code, rather than printing a proper diagnostic.

None of the tests are particularly long, but here's an overview of the failure cases (caveat: I haven't investigated most of these, so take these descriptions as educated guesses at the root cause):

1. Importing an inlinable function body that contains `new` of an instantiated type crashes.
2. Recursive instantiation of a generic method crashes during dictionary insertion.
3. Importer crashes on channel types (and probably others) embedded into a type parameter constraint.
4. Importer/exporter don't support labeled for loops, so instantiating an imported generic function that contains one crashes during SSA because the label references are unresolved.
5. Inconsistent mangling when instantiating a type with anonymous interface as a type argument: `X[interface{}]` vs `X[interface {}]`.
6. Method expressions involving an instantiated interface type fail.
7. Importer crashes on importing an instantiated interface.
8. Embedded `comparable` constraint is lost during -G=3 noding, so consequently it isn't exported or re-imported. This causes incorrect code to be accepted, and could also (potentially) cause correct code to be misoptimized (e.g., assuming that simple GC-shape stenciling is appropriate, rather distinguishing types that require different equality operations, like `uint32` and `float32` or `string` and `struct { *byte; int }`).
9. Exporter does not include intermediary types for composite literal keyed elements.
10. Method expression involving an imported instantiated interface types result in a linker error because of a missing wrapper function. (Note: currently masked by issue 6.)
11. Compiler crash on invalid code involving `//go:notinheap` rather than an error message.

# Issue 10 and Unified IR

I knew one backend feature I needed for unified IR was to create type descriptors for imported generic types and mark them as DUPOK. While looking at the reflectdata code to make this happen, I saw a `Type.IsFullyInstantiated` flag had been added for -G=3 and was used to handle this already. I also saw other code checked this. Rather than duplicate this logic, I thought it would be less invasive to arrange for `Type.IsFullyInstantiated` to report true for unified IR's instantiated types too.

However, as issue 10 demonstrates, the existing support for that flag is incomplete. In particular, reflectdata.methodWrapper was extended for -G=3 to generate method wrappers for imported generic concrete types, but not to generate method wrappers for imported generic interface types:

https://github.com/golang/go/blob/b20747334a4a3dee51759369a098ef2a0c9dbcff/src/cmd/compile/internal/reflectdata/reflect.go#L1794-L1806

# Future work

I think fuzz testing would be very useful (/cc @thepudds). I think there's a lot of room for testing more complex language features (e.g., function literals; go, defer, range, and select statements; multi-package tests with re-exported generic types and functions; tests with identically named packages).

I also think as dictionary and GC shape support continues, actually fuzz testing the resulting code (i.e., as opposed to just verifying the compiler doesn't crash) will be important too.

/cc @danscales @randall77 @griesemer @ianlancetaylor @rsc @eliben

**Comment by cuonglm** [comment]
After #46725 fixed, `test/fixedbugs/issue46725.go` is also failed with `-G=3`

**Comment by cuonglm** [comment]
Another failure case:

```
$ cat go.mod
module foo

go 1.18

$ cat a/a.go
package a

type S[T any] struct {
        F T
}

var X = S[int]{}

$ cat main.go
package main

import (
	"foo/a"
)

func main() {
	_ = a.X
}

$ go run -gcflags='all=-G=3' main.go
# command-line-arguments
./a/a.go:3:6: internal compiler error: failed to find type loop for: a.S[int]
<truncated stack trace>
```

**Comment by mdempsky** [comment]
@cuonglm That looks pretty similar to test case 1's failure. Maybe the issue is related to the blank assignment, not the `new` call like I initially conjectured.

**Comment by cuonglm** [comment]
@mdempsky it seems to me iimport.go:L707 missed the condition for fully instantiated type.

**Comment by mdempsky** [comment]
Another test program that crashes with -G=3 (and go2go), but passes with unified IR. This one fails with "mer.go:14:5: internal compiler error: Unexpected op with CALL during stenciling: METHEXPR" when compiled with -G=3:

```
package main

type Mer interface{ M() }

func F[T Mer](expectPanic bool) {
	defer func() {
		err := recover()
		if (err != nil) != expectPanic {
			print("FAIL: (", err, " != nil) != ", expectPanic, "\n")
		}
	}()

	var t T
	T.M(t)
}

type MyMer int

func (MyMer) M() {}

func main() {
	F[Mer](true)
	F[MyMer](false)
	F[*MyMer](true)
	F[struct{ MyMer }](false)
	F[struct{ *MyMer }](true)
	F[*struct{ MyMer }](true)
	F[*struct{ *MyMer }](true)
}
```

**Comment by mdempsky** [comment]
This program fails to compile with -G=3 due to "internal compiler error: 'main': Value live at entry. It shouldn't be. func main, node .dict0, value nil":

```
package main

type T[_ any] int
func (T[_]) M() {}

type I interface{ M() }

type U = T[int]
var x = U(0)
var i I = x

func main() {
	U.M(x)
	i.M()
}
```

**Comment by danscales** [comment]
Thanks for reporting and making great examples for the bugs, Matthew and Cuong!  Here's an update:

I diagnosed all of the bugs fairly easily (in about 90 minutes), and also created fixes for most of them.  I'll be making CLs with the fixes over the next few days.  Two of them are already out for review.

A couple notes:

- 6.go and 10.go are already fixed - I believe that they were both dictionary-related bugs.  We are actively developing dictionaries and unified IR does not include our dictionary analyzing/transforming code.  14.go is also a known dictionary issue and I have a fix out for that.
- 3.go is just relating to switching the default from 1.17 to 1.18
- 4.go is the known bug that we don't export/import functions with labeled break statements.  (I know that unified IR can already do this.)  This is the only bug that I won't immediately fix.
- 11.go appears to be a bug in types2 (types2 is not finding a mismatch between two types which only differ in having/not having //go:notinheap).  This bug doesn't show up in unified IR, because unified IR is running both types2 and types1 typechecking on most code.  -G=3 uses only types2 typechecking for most user code.   I'll report that to Robert.

**Comment by cuonglm** [comment]
> 11.go appears to be a bug in types2 (types2 is not finding a mismatch between two types which only differ in having/not having //go:notinheap). This bug doesn't show up in unified IR, because unified IR is running both types2 and types1 typechecking on most code. -G=3 uses only types2 typechecking for most user code. I'll report that to Robert.

I think this is not a bug in types2, as pragma is not part of the language spec, and it exists for the compiler, not for the typechecker.

**Comment by danscales** [comment]
> > 11.go appears to be a bug in types2 (types2 is not finding a mismatch between two types which only differ in having/not having //go:notinheap). This bug doesn't show up in unified IR, because unified IR is running both types2 and types1 typechecking on most code. -G=3 uses only types2 typechecking for most user code. I'll report that to Robert.
> 
> I think this is not a bug in types2, as pragma is not part of the language spec, and it exists for the compiler, not for the typechecker.

OK, thanks, Robert confirmed that he does not currently deal with pragmas in types2 and is hoping to keep it that way.  I will try to handle this in noder2.

**Comment by anacrolix** [comment]
I tried to find the right place to report this, and kept finding my way to this issue:
```
# github.com/anacrolix/stm [github.com/anacrolix/stm.test]
./funcs.go:148:30: internal compiler error: Unexpected op with CALL during stenciling: function call

goroutine 1 [running]:
runtime/debug.Stack()
	/Users/anacrolix/src/go.master/src/runtime/debug/stack.go:24 +0x65
cmd/compile/internal/base.FatalfAt({0x190b98d, 0x0}, {0xc000bbf740, 0x38}, {0x0, 0x0, 0x0})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/base/print.go:227 +0x154
cmd/compile/internal/noder.(*subster).node.func1({0x1a62198, 0xc000aef290})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:1104 +0x119a
cmd/compile/internal/ir.editNodes({0xc000c2a7a0, 0x1, 0x18ce320}, 0xc000c32510)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:1521 +0x74
cmd/compile/internal/ir.(*ReturnStmt).editChildren(0xc000c31c20, 0x1a63cf0)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:1056 +0x53
cmd/compile/internal/ir.EditChildren(...)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/visit.go:185
cmd/compile/internal/noder.(*subster).node.func1({0x1a63cf0, 0xc000afcf00})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:942 +0x486
cmd/compile/internal/ir.editNodes({0xc000c3a2a0, 0x2, 0x18cb080}, 0xc000c32510)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:1521 +0x74
cmd/compile/internal/ir.(*IfStmt).editChildren(0xc000c3e150, 0xc000c32510)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:654 +0xb4
cmd/compile/internal/ir.EditChildren(...)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/visit.go:185
cmd/compile/internal/noder.(*subster).node.func1({0x1a62e18, 0xc000b047e0})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:942 +0x486
cmd/compile/internal/ir.editNodes({0xc000c31810, 0x5, 0xc000c38501}, 0xc000c32510)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:1521 +0x74
cmd/compile/internal/ir.(*CaseClause).editChildren(0xc000c2f6e0, 0xc000c32510)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:309 +0xba
cmd/compile/internal/ir.EditChildren(...)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/visit.go:185
cmd/compile/internal/noder.(*subster).node.func1({0x1a62260, 0xc000af3e60})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:942 +0x486
cmd/compile/internal/ir.editCaseClauses({0xc000c32528, 0x3, 0x18cbb20}, 0xc000c32510)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:1449 +0x62
cmd/compile/internal/ir.(*SwitchStmt).editChildren(0xc000c11880, 0xc000c32510)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:1320 +0xb4
cmd/compile/internal/ir.EditChildren(...)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/visit.go:185
cmd/compile/internal/noder.(*subster).node.func1({0x1a644c0, 0xc000728b80})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:942 +0x486
cmd/compile/internal/noder.(*subster).node(0xc000c11800, {0x1a644c0, 0xc000728b80})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:1248 +0xa5
cmd/compile/internal/noder.(*subster).list(0x18729e0, {0xc000722d00, 0x1, 0xc00052ad10})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:1362 +0x8e
cmd/compile/internal/noder.(*subster).node.func1({0x1a623f0, 0xc000afc280})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:1166 +0x1a45
cmd/compile/internal/ir.editNodes({0xc000c2a640, 0x1, 0x18ce320}, 0xc000c32468)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:1521 +0x74
cmd/compile/internal/ir.(*ReturnStmt).editChildren(0xc000c314a0, 0x1a63cf0)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:1056 +0x53
cmd/compile/internal/ir.EditChildren(...)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/visit.go:185
cmd/compile/internal/noder.(*subster).node.func1({0x1a63cf0, 0xc000afcfa0})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:942 +0x486
cmd/compile/internal/noder.(*subster).node(0xc000c11800, {0x1a63cf0, 0xc000afcfa0})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:1248 +0xa5
cmd/compile/internal/noder.(*subster).list(0xc000610148, {0xc000722df0, 0x1, 0xc0001345f0})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:1362 +0x8e
cmd/compile/internal/noder.(*irgen).genericSubst(0xc00050e400, 0xc000c31310, 0xc0000d9c70, {0xc0006100e0, 0x1, 0x1}, 0x0, 0xc000c34740)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:726 +0xce6
cmd/compile/internal/noder.(*irgen).getInstantiation(0xc00050e400, 0xc0000d9c70, {0xc0006100d8, 0x1, 0x1}, 0x98)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:608 +0x33d
cmd/compile/internal/noder.(*irgen).stencil.func1({0x1a62198, 0xc000bb61b0})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:109 +0x2df
cmd/compile/internal/ir.Visit.func1({0x1a62198, 0xc000bb61b0})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/visit.go:105 +0x30
cmd/compile/internal/ir.doNodes({0xc000c0bf40, 0x2, 0xc000bac420}, 0xc000c32390)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:1512 +0x67
cmd/compile/internal/ir.(*CallExpr).doChildren(0xc000bb6240, 0xc000c32390)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:263 +0x85
cmd/compile/internal/ir.DoChildren(...)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/visit.go:94
cmd/compile/internal/ir.Visit.func1({0x1a62198, 0xc000bb6240})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/visit.go:106 +0x57
cmd/compile/internal/ir.doNodes({0xc000b81740, 0x1, 0x0}, 0xc000c32390)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/node_gen.go:1512 +0x67
cmd/compile/internal/ir.(*Func).doChildren(0x1a62af8, 0xc000bac420)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/func.go:152 +0x2e
cmd/compile/internal/ir.DoChildren(...)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/visit.go:94
cmd/compile/internal/ir.Visit.func1({0x1a62af8, 0xc000bac420})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/visit.go:106 +0x57
cmd/compile/internal/ir.Visit({0x1a62af8, 0xc000bac420}, 0xc000c346c0)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/ir/visit.go:108 +0xb8
cmd/compile/internal/noder.(*irgen).stencil(0xc00050e400)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/stencil.go:91 +0x26a
cmd/compile/internal/noder.(*irgen).generate(0xc00050e400, {0xc000433a90, 0x9, 0x0})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/irgen.go:304 +0x359
cmd/compile/internal/noder.check2({0xc000433a90, 0x9, 0x9})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/irgen.go:92 +0x175
cmd/compile/internal/noder.LoadPackage({0xc0001320e0, 0x9, 0x0})
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/noder/noder.go:90 +0x335
cmd/compile/internal/gc.Main(0x191aec0)
	/Users/anacrolix/src/go.master/src/cmd/compile/internal/gc/main.go:190 +0xaf3
main.main()
	/Users/anacrolix/src/go.master/src/cmd/compile/main.go:55 +0xdd

FAIL	github.com/anacrolix/stm [build failed]
FAIL
```
Running `go test .` on https://github.com/anacrolix/stm/commit/80e033aa1f2218b83fb5891671ed795de72e19d5. Running with `GOEXPERIMENT=unified go test .` I get instead:
```
$ 
# github.com/anacrolix/stm/stmutil
vet: stmutil/context.go:7:2: could not import github.com/anacrolix/stm (cannot import "github.com/anacrolix/stm" (runtime error: invalid memory address or nil pointer dereference), possibly version skew - reinstall package)
# github.com/anacrolix/stm_test
vet: ./external_test.go:8:2: could not import github.com/anacrolix/stm (cannot import "github.com/anacrolix/stm" (runtime error: invalid memory address or nil pointer dereference), possibly version skew - reinstall package)
FAIL	github.com/anacrolix/stm [build failed]
```

go version devel go1.18-4957976b1a Fri Sep 3 22:56:58 2021 +0000 darwin/amd64

**Comment by cuonglm** [comment]
@anacrolix You need `GOEXPERIMENT=unified go test -vet=off .` for now.

**Comment by danscales** [comment]
@anacrolix   Your problem with stm.test is fixed by the fix of #48042 
 https://github.com/golang/go/commit/ecfff58fb8c3aabbce7b15c850210485c1f09d61 , just checked in.

If you have any other problem with type params, please file a new issue mentioning  -G=3 mode.  Thanks!

**Comment by anacrolix** [comment]
Thanks!

**Comment by danscales** [comment]
Closing this bug, now that all the issues mentioned fixed.  The last one was 4.go, fixed by https://go-review.googlesource.com/c/go/+/349012 (and following fix to ORANGE import/export https://go-review.googlesource.com/c/go/+/350911).
